### PR TITLE
Add operator position to call expressions

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Add operator position to call (FunctionCallExpr, MethodCallExpr, CheckingExpr, AST expressions
Add operator position to corresponding BIR instructions
Related to #604, #605

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front.syntax/ast.bal [hunk 1]
```diff
@@ -223,6 +223,7 @@ public type FunctionCallExpr record {|
     // *PositionFields
     Position startPos;
     Position endPos;
+    Position opPos;
     string? prefix = ();
     string funcName;
     Expr[] args;
```
Comment (heshanpadmasiri) [inline compiler/modules/front.syntax/ast.bal:229]: @manuranga, @jclark when I remove `pos` below I get a java exception as fallows:
```
[2021-11-01 09:24:55,605] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkFieldAccessExpr(TypeChecker.java:7114)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:2562)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangFieldBasedAccess.accept(BLangFieldBasedAccess.java:65)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:374)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:353)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:349)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.setTypeOfVarRef(SemanticAnalyzer.java:4021)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.setTypeOfVarRefInAssignment(SemanticAnalyzer.java:4015)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1852)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangAssignment.accept(BLangAssignment.java:80)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.lambda$visit$28(SemanticAnalyzer.java:1739)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1739)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:59)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:2346)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangIf.accept(BLangIf.java:56)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.lambda$visit$28(SemanticAnalyzer.java:1739)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1739)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:59)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:2346)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangIf.accept(BLangIf.java:56)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:477)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:447)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3766)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:333)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.lambda$visit$2(SemanticAnalyzer.java:344)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:344)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:163)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:299)
        at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:206)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:114)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:403)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:349)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:201)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:186)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:111)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:100)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:214)
        at io.ballerina.projects.Package.getCompilation(Package.java:140)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:75)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:401)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
ERROR [nballerina.front.syntax:(1:1,1:1)] Compilation failed due to an unhandled exception
```
(same happens for `MethodCallExpr` as well) any suggestion on what might cause this?

File: compiler/modules/front.syntax/ast.bal [hunk 2]
```diff
@@ -248,6 +250,7 @@ public type CheckingExpr record {|
     // *PositionFields
     Position startPos;
     Position endPos;
+    Position opPos;
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: Is this the position of the `check` keyword? If so, then `opPos` is not a good name.  Use the same word here as for `error`. Maybe `kwPos`.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/ast.bal (hunk 1 above)
Comment (manuranga) [reply compiler/modules/front.syntax/ast.bal:229]: Just for the record, we speculate that this is due to some references to pos `field`, and jBallerina is crashing instead giving `field access to non-required member`.
Let's 
1. Remove all such references and reattempt.
2. Create a issue in jballerina when you narrow down the minimal case

Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal:229]: created [issue](https://github.com/ballerina-platform/ballerina-lang/issues/33566)

File: compiler/modules/front.syntax/ast.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal]: fixed

File: compiler/modules/front.syntax/tests/testPos.bal [hunk 3]
```diff
@@ -301,26 +302,38 @@ function validateExpressionPos(Expr expr, Tokenizer tok, Position parentStartPos
...
         }
         else if expr is BinaryArithmeticExpr {
-            test:assertTrue(opToken is BinaryArithmeticOp);
+            test:assertTrue(token is BinaryArithmeticOp);
         }
         else if expr is BinaryBitwiseExpr {
-            test:assertTrue(opToken is BinaryBitwiseOp);
+            test:assertTrue(token is BinaryBitwiseOp);
+        }
+        else if expr is UnaryExpr{
```
Comment (jclark) [inline compiler/modules/front.syntax/tests/testPos.bal:329]: Space before brace

File: compiler/modules/front.syntax/ast.bal [hunk 4]
```diff
@@ -259,6 +259,7 @@ public type CheckingStmt record {|
     CheckingKeyword checkingKeyword;
     Position startPos;
     Position endPos;
+    Position kwPos;
     CallStmt operand;
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal:263]: Not relevant to this review but CheckingStmt should contain a CallExpr

File: compiler/modules/front.syntax/parseStmt.bal [hunk 5]
```diff
@@ -221,18 +220,19 @@ function finishCallStmt(Tokenizer tok, CallStmt expr, Position startPos) returns
 function finishCheckingCallStmt(Tokenizer tok, CheckingKeyword checkingKeyword, Position startPos) returns CallStmt|err:Syntax {
     Token? t = tok.current();
     if t is "check"|"checkpanic" {
+        Position operandStartPos = tok.currentStartPos();
```
Comment (jclark) [inline compiler/modules/front.syntax/parseStmt.bal:223]: Don't understand what's going on here. The current token start position is the keyword position, but you're getting the keyword positionm from somewhere else.

File: compiler/modules/front.syntax/parseStmt.bal [hunk 6]
```diff
@@ -221,18 +220,19 @@ function finishCallStmt(Tokenizer tok, CallStmt expr, Position startPos) returns
 function finishCheckingCallStmt(Tokenizer tok, CheckingKeyword checkingKeyword, Position startPos) returns CallStmt|err:Syntax {
```
Comment (jclark) [inline compiler/modules/front.syntax/parseStmt.bal:220]: What is startPos the position of here? This needs a comment.

File: compiler/modules/front.syntax/parseStmt.bal [hunk 7]
```diff
@@ -221,18 +220,19 @@ function finishCallStmt(Tokenizer tok, CallStmt expr, Position startPos) returns
...
     else if t == "(" {
         MethodCallExpr operand = check parseMethodCallStmt(tok);
-        return { startPos, endPos: operand.endPos, checkingKeyword, operand };
+        return { startPos, endPos: operand.endPos, kwPos: startPos, checkingKeyword, operand };
     }
     Expr operand = check parsePrimaryExpr(tok);
     if operand is FunctionCallExpr|MethodCallExpr {
         Position endPos = check tok.expectEnd(";");
-        return { startPos, endPos, checkingKeyword, operand };
+        return { startPos, endPos, kwPos: startPos, checkingKeyword, operand };
```
Comment (jclark) [inline compiler/modules/front.syntax/parseStmt.bal:235]: Why is kwPos not coming from the position of the "check" or "checkpanic" token?

**Review by jclark — CHANGES_REQUESTED** [review]

**Comment by heshanpadmasiri** [comment]
closed in favor of #663

File: compiler/modules/front.syntax/parseStmt.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/parseStmt.bal:223]: this function is called after ingesting `CheckingKeyword` here we are dealing with the case like `check check f()` and `operandStartPos` is the `startPos` of second `check`

File: compiler/modules/front.syntax/parseStmt.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/parseStmt.bal:220]: `startPos` is the position of the `checkingKeyword` (parameter)

File: compiler/modules/front.syntax/parseStmt.bal (hunk 7 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/parseStmt.bal:235]: startPos is the position of the checkingKeyword (parameter)

File: compiler/modules/front.syntax/parseStmt.bal (hunk 6 above)
Comment (jclark) [reply compiler/modules/front.syntax/parseStmt.bal:220]: Then call it keywordPos or kwPos.

File: compiler/modules/front.syntax/parseStmt.bal (hunk 5 above)
Comment (jclark) [reply compiler/modules/front.syntax/parseStmt.bal:223]: Then call it innerKwPos.

**Pull request closed without merging**